pickeduname patch cumulative v4 - 1st git commit from diff-v4

pickeduname opened the pull request:
@MCMrARM I git-push-ed my v4 diff, this is the pull request for it.
- all new features as discussed in thread #295 and documented in `README.md`, this commit bumps `versionName -> "0.5.4f"` and `versionCode -> 18`. Upon merge, perhaps bump these further.
- _Please_ never decrement versionCode for any reason. The next release version code should then be _at least_ 19 and version name "0.5.5" _imnsho_ .
- patched git cloned repo today (9 May 2021), no changes from v4 diff which is verified to build and yield a working debug APK on Oreo (8.1). _Patched tree in git verified to build debug APK with no errors_
- just one diff v4 fuzz found - patch message about one line offset in `strings.xml`, leaves `strings.xml.orig` in values dir, that .orig must be deleted manually before building in AS. The step is only needed when patching the cloned repo with patch, not after git this commit. The patch v4 is in #295, in my last message \@.
- I know of no bugs in the code submitted. If there are any, I want to know about them :smile:
- One thing which will be done some time: change some vars in the patch from `Integer` to `int`. Not urgent.

MCMrARM commented on app/src/main/java/io/mrarm/irc/ServerConnectionInfo.java:
```diff
@@ -129,12 +131,29 @@ public void connect() {
+                    appname = mManager.getContext().getString(R.string.app_name);
+                    appver = BuildConfig.VERSION_NAME;
+                }
+                messageHandler.setCtcpVersionReply(appname, appver, "Android");
```
This is an unrelated change

MCMrARM commented on app/src/main/java/io/mrarm/irc/ServerConnectionInfo.java:
```diff
@@ -129,12 +131,29 @@ public void connect() {
+                String errstr = ctx.getString(R.string.notification_server_connection_info_badcount, mServerConfig.badCount);
+                Toast.makeText(ctx, errstr, Toast.LENGTH_LONG).show();
+                Log.d("ServerConnectionInfo", errstr);
+            }
```
I don't think it makes sense to show this error every app startup?

MCMrARM commented on app/build.gradle:
```diff
@@ -7,8 +7,8 @@ android {
         targetSdkVersion 30
-        versionCode 12
-        versionName "0.5.4"
+        versionCode 18
```
Please do not change the version numbers in a pull request; pull requests are merged to the master branch and do not mean a new release; the code will be tagged and versioned manually.

MCMrARM commented on README.md:
```diff
@@ -8,6 +8,30 @@ Revolution IRC Client is the next-generation IRC client for Android, made with d
 <a href="https://play.google.com/store/apps/details?id=io.mrarm.irc" target="_blank">
 <img src="https://play.google.com/intl/en_us/badges/images/generic/en-play-badge.png" alt="Get it on Google Play" height="80"/></a>
 
+---
```
This should be removed from here if this is to be merged.

MCMrARM commented on app/src/main/java/io/mrarm/irc/IgnoreListAdapter.java:
```diff
@@ -115,6 +125,17 @@ public void bind(ServerConfigData.IgnoreEntry entry) {
             }
+
+            if (entry.mesg != null) {
+                if (builder.getSpannable().length() != 0) builder.append(" " );
```
As I noted in the other pull request, this should be done in the XML using android:ellipsize="end" android:maxLines="1"

MCMrARM commented on app/src/main/java/io/mrarm/irc/IgnoreListAdapter.java:
```diff
@@ -94,6 +100,10 @@ private void startEdit() {
         public void bind(ServerConfigData.IgnoreEntry entry) {
             ColoredTextBuilder builder = new ColoredTextBuilder();
+            if (entry.isBad) {
+                builder.append("@bad:", new ForegroundColorSpan(Color.RED), new StyleSpan(Typeface.BOLD));
```
This should probably use an android string for the prefix? Also it should probably read `"Bad: "` not `"@bad:"`

MCMrARM commented on app/src/main/java/io/mrarm/irc/util/IgnoreListMessageFilter.java:
```diff
@@ -13,20 +18,43 @@
+        mConfig.badCount = checkCompileIgnoreListEntries();
+    }
+
+    private Integer checkCompileIgnoreListEntries() {
```
should return `int`; i do not think it needs to return anything if the warning toast is removed though?

MCMrARM commented on app/src/main/java/io/mrarm/irc/util/SimpleWildcardPattern.java:
```diff
@@ -25,7 +26,29 @@ public static Pattern compile(String str) {
+        return ret.toString();
     }
 
+    public static Pattern pattCompile(String str) throws PatternSyntaxException {
```
Please apply the suggestion made by me; code like this is unacceptable in Java.

MCMrARM commented:
Let me reply to the notes from the other PR:

> 1. reverting version numbers and text in `README.md` and `app/build.gradle` can be done but I prefer to keep those as I pasted them. A merge-clean branch with these reverted will be made if requested. But I usually bump versionCode and Name liberally to keep changes visible. _Please never ever decrement version numbers for any reason_. Reverting "my" version number used for debugging is perfectly fine on merge, of course.
as I said, it's my responsibility to increment the version number with the next release; PR have no relationship with releases so this is the wrong place to do an increment; README.md is the project readme; the notes about what the PR changes should be in the PR description
> 2. bracketing re's with `//` globs with `??` or with _no brackets_ (default), seems to work well. The hint strings in the Ignore Entry screen have been edited so they show _Any /regexp/ or ?glob?_, i.e. no surprise to the user.
This is okay for now, althrough not the best UX
> 3. the way `pattCompile()` is written is the _C way_, as I intend to use the code elsewhere too. The goal is to set up the call to the real pattern compile function only once, since that is the expensive call usually. The `int j` is used to offset / cut the head and tail of the string, thus values 0 or 1 and it is not a boolean for the same reason. Could have named it `i` for index, `j` is also used as an index integer frequently. Initially I used `i` for `isRe` but then renamed it and it became a boolean. This is how it stayed `j` and definitely int. The way @MCMrARM coded it in #296 is a more Java esque idiom and terser, I think his version is better, for Java. Similarly, the `compile()` function name is imo not reflecting the true nature of the function. But it could be reverted, or ... see 9. below.
The project is Java. Easy code reuse is no excuse not to follow a language's idioms.
> 4. `isBad` is not transient and is saved in the `.zip` exported settings and in persistent settings. Note there's a transient `badCount` too, which is used to pass the count of bad entries from the compile / check function back up to the caller which can raise the toast. `badCount` counts the number of set `isBad` flags / entries. `isBad` is used to mark the Ignore List entries `@bad:` in the ListAdapter among other things, which could (should?) be available for edit before connecting to a server, but isn't currently so, hope it will be some day. This means that the list adapter is able to display the updated marked bad re containing entries without calling the check/compile function itself. Note that the syntax check is triggered from 2 places now: 1) from `connect()` and 2) from exit from an Ignore Entry edit. The latter only checks the current screen's re's, not all, the former checks all. Without a persistent `isBad`, it would have been necessary for the list adapter to call the compile / check function.
I think it should be fine to force recompilation of bad regexes every app launch though, since we already need to recompile the valid regexes and it's pointless to optimize for an error case. I would mark `isBad` as volatile here.
> 5. `isBad` and `badCount` are both `Integer` but could be `int` and might become `int`'s.
Do not use Integer if possible since that's a boxed type, always use `int` unless you can't.
> 6. Ellipsizing the Ignore Entry text is indeed not using the facilities provided by Android but I would like there to be at least a few characters of the text match pattern in the line, no matter how _small_ the screen is. I have devices with very small screens.
Post a screenshot illustrating this issue; either way no, trimming the text in code is not okay. I do not think this issue would be limited to this screen only though.
> 7. Similarly, the count of bad re's is not pluralized properly using the mechanisms provided by Android strings.xml etc., this is seen in ` app/src/main/java/io/mrarm/irc/ServerConnectionInfo.java` @~L153.
I don't think displaying the count on app startup helps the user in any way; I think it's just adding unneeded noise and nagging. This will also possibly display on transparent app restarts in the background; I think it's best __not__ to bother the user with an unneeded check on app startup since they already have got the error when editing the regex.
> 8. re: `app/src/main/java/io/mrarm/irc/EditIgnoreEntryActivity.java` @~L95: Indeed that toast could use Android strings and with that also the pluralization mechanism (6. above).
This should only display an error related to the currently editing entry removing the need for a plural string.

pickeduname commented:
Re: code changes: will have to wait for v5 patch, Sunday probably. @MCMrARM Your comments are pertinent and I reacted to them, in v5, which you have not seen yet (Sunday hopefully).

All the issues you brought up in #296 were addressed, strings, etc. There are 2 or 3 things I left as is for now but may be done by Sunday. One is ellipsization (sp?). All strings are in strings.xml and I refactored / moved classes as you suggested, moved the strange decision code in the regexp/glob compiler to your version suggested in #296 etc.

I will revert the versioning for the branch for which the pull request will go out (v5), but I need to keep my own branch on different version numbers else I have trouble keeping track of on-device installed versions.

We'll see by Sunday night (my time is Eastern EU TZ).

pickeduname commented:
@MCMrARM :

While editing the v5 patch I found that AS deletes two lines from `config/ServerConfigData.java`, which are:

- import java.security.spec.EncodedKeySpec;

- import io.mrarm.irc.R;

--- it does this quietly without asking me and is consistent, it also did it in AS 4.0 before I upgraded to 4.2 .

Do you know what causes this? Can I leave it in the patch as is? I do not see a way to fix it for now, other than manually editing the sources, to put the lines back, with AS closed.
